common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/constraints"
+
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/dynamicconfig"
+)
+
+type (
+	// history holds the running per-second running average for a single key from a single host, and the last time it was updated.
+	history struct {
+		lastUpdate         time.Time // only so we know if elapsed times are exceeded, not used to compute per-second rates
+		accepted, rejected PerSecond // requests received, per second (conceptually divided by update rate)
+	}
+
+	// Identity is an arbitrary (stable) identifier for a "limiting" host.
+	Identity string
+	// Limit is the key being used to Limit requests.
+	Limit string
+	// PerSecond represents already-scaled-to-per-second RPS values
+	PerSecond float64
+	// HostWeight is a float between 0 and 1, showing how much of a ratelimit should be allocated to a host.
+	HostWeight float64
+
+	impl struct {
+		cfg Config
+
+		// mut protects usage, as it is the only mutable data
+		mut sync.Mutex
+		// usage data for ratelimits, per host.
+		//
+		// data is first keyed on the limit and then on the host, as it's assumed that there will be
+		// many times more limit keys than limiting hosts, and this will reduce cardinality more rapidly.
+		usage map[Limit]map[Identity]history
+
+		clock clock.TimeSource
+	}
+
+	// Metrics reports overall counts discovered as part of garbage collection.
+	//
+	// This is not necessarily worth maintaining verbatim, but while it's easy to
+	// collect it might give us some insight into overall behavior.
+	Metrics struct {
+		// HostLimits is the number of per-host limits that remain after cleanup
+		HostLimits int
+		// Limits is the number of cross-host limits that remain after cleanup
+		Limits int
+
+		// RemovedHostLimits is the number of per-host limits that were removed as part of this GC pass
+		RemovedHostLimits int
+		// RemovedLimits is the number of cross-host limits that were removed as part of this GC pass
+		RemovedLimits int
+	}
+
+	// Requests contains accepted/rejected request counts for a ratelimit, as part of an update operation.
+	// Only intended to be used in [RequestWeighted.Update].
+	Requests struct {
+		Accepted, Rejected int
+	}
+
+	// RequestWeighted returns aggregated RPS and distribution data for its Limit keys,
+	// based on how many requests each Identity has received.
+	RequestWeighted interface {
+		// Update load-data for this host's requests, given a known elapsed time spent accumulating this load info.
+		//
+		// Elapsed time must be non-zero, but is not otherwise constrained.
+		Update(host Identity, load map[Limit]Requests, elapsed time.Duration)
+
+		// HostWeights returns the per-[Limit] weights for all requested + known keys for this Identity,
+		// as well as the Limit's overall used RPS (to decide RPS to allow for new hosts).
+		HostWeights(host Identity, limits []Limit) (weights map[Limit]HostWeight, usedRPS map[Limit]PerSecond)
```

Comment:
is host weight used to calculate the limit for that particular host? and is it calculated based on history requests?

Reply:
yea, I think I'll draw an ascii diagram of the whole thing for the docs.  it should help clarify stuff.

for the meantime tho:
- normal calls `Update` to push the data in, then get their `HostWeights` to get the info needed to compute RPS to respond
- yes, changing hosts will change identities, and they'll be empty until they update.  we currently have no way of knowing if they'll actually replace [host X] in a request-routing role so we can't really take advantage of previous data.  that's fine, it's just the empty state which we need to handle anyway.
  - we could return all known keys for an "at startup" request to mitigate this - I've got that built elsewhere actually.  I'm not sure if it's worth doing though, as we still have to handle delays (either in get-rps-response or in accept-requests)
  - alternatively, it can just fall back to the in-memory `RPS/hosts` count for one update cycle.  we need this anyway for safety.
  - this is all also true for aggregating-hosts disappearing or moving, we lose a shard of data like all our other caches.  limiting hosts will keep their last-received rps temporarily to cover these gaps.

Reply:
added quite a lot of global/doc.go and docs here - lemme know if that clears this up :)